### ISSUE TITLE
Feature request - visualize uncertainty

### ISSUE THREAD

**Issue by Eugenio-Bruno** [kiudee/chess-tuning-tools#94]
I'm pretty sure I don't understand how this works under the hood, so this might not make sense...

If I understand correctly (big if), for each possible parameters (p_1, p_2, ..., p_n) the optimizer has both an estimate of the value and an uncertainty regarding the value at those parameters.

Would it be possible to have plots that not only show the estimate of the value, but also the uncertainty over the space it's searching? Maybe, like there are partial dependence plots, there could be "partial uncertainty" plots, and like there are the 2d plots there could be uncertainty plots for combinations of two parameters.

The reason this might be useful is to see how far the optimizer has gotten and how well it thinks it has explored the space.

Again, I don't know if this even makes sense - feel free to just close the issue if it doesn't :p

**Comment by Eugenio-Bruno** [comment]
This might also help me understand the tradeoff between number of iterations and rounds per iteration...

I've been trying vanilla cutechess-cli tournaments with an engine I'm trying to tune, with the same parameters and time control, which should give more or less 0 elo, but even at 4000 matches, I still get stuff like `Elo difference: 8.34 +/- 8.59`, which makes me feel like I do not understand this tradeoff at all. At 4000 matches I'd expect very little noise but that looks like a ton - maybe not much of an improvement from say 1000 matches.

By the way, when I do one iteration with 50 games, does the optimizer get 50 noisier {0, 0.5, 1} results, or one single less noisy result in [0, 1]?

**Comment by kiudee** [comment]
> I've been trying vanilla cutechess-cli tournaments with an engine I'm trying to tune, with the same parameters and time control, which should give more or less 0 elo, but even at 4000 matches, I still get stuff like `Elo difference: 8.34 +/- 8.59`, which makes me feel like I do not understand this tradeoff at all. At 4000 matches I'd expect very little noise but that looks like a ton - maybe not much of an improvement from say 1000 matches.

Yes, sadly this is realistic. That is the reason why on [Fishtest](https://tests.stockfishchess.org) they have to run up to 100k games to be able to accurately determine Elo gaining patches, 
> 
> By the way, when I do one iteration with 50 games, does the optimizer get 50 noisier {0, 0.5, 1} results, or one single less noisy result in [0, 1]?

The latter. It will aggregate the results in a pentanomial model with counts for the five classes (WW, WD, WL/DD, DL, LL). The win probability will be converted into an Elo number in ℝ.

**Comment by kiudee** [comment]
> If I understand correctly (big if), for each possible parameters (p_1, p_2, ..., p_n) the optimizer has both an estimate of the value and an uncertainty regarding the value at those parameters.

That is correct.
> 
> Would it be possible to have plots that not only show the estimate of the value, but also the uncertainty over the space it's searching? Maybe, like there are partial dependence plots, there could be "partial uncertainty" plots, and like there are the 2d plots there could be uncertainty plots for combinations of two parameters.
> 
> The reason this might be useful is to see how far the optimizer has gotten and how well it thinks it has explored the space.

That is doable. I think to be useful it would also need a good scale bar where the deviation in Elo points is visible.
